### PR TITLE
Feature/error reviver

### LINKED ISSUE #64: Implement `error-reviver`
See [utils-error-reviver](https://github.com/kgryte/utils-error-reviver). `@stdlib/utils/error-reviver`.

### PR DESCRIPTION
**Pull request by Planeshifter**
Resolves #64 .

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

* [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
* [x] Read and understand the [Code of Conduct][code-of-conduct].
* [x] Read and understood the [licensing terms][license].
* [x] Searched for existing issues and pull requests __before__ submitting this pull request.
* [x] Filed an issue (or an issue already existed) __prior to__ submitting this pull request.
* [x] Rebased onto latest `develop`.
* [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

* contains a straight-forward port of [utils-error-reviver](https://github.com/kgryte/utils-error-reviver)

## Related Issues

> Does this pull request have any related issues?

This pull request:

* resolves #
* fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

---

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/master/CONTRIBUTING.md
[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/master/CODE_OF_CONDUCT.md
[license]: https://github.com/stdlib-js/stdlib/blob/master/LICENSE

### CODE REVIEW

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
+# reviveError
+
+> Revive a JSON-serialized [error][utils-error-to-json] object.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: `@stdlib/utils/error-to-json`

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+## Usage
+
+``` javascript
+var reviveError = require( '@stdlib/utils/error-reviver' );
+```
+
+#### reviveError( key, value )
+
+Revives a JSON-serialized [error][utils-error-to-json] object.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: `@stdlib/utils/error-to-json`

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+``` javascript
+var str = '{"type":"TypeError","message":"beep"}';
+
+var err = JSON.parse( str, reviveError );
+// returns <TypeError>
+```
+
+For details on the JSON serialization format, see [error-to-json][utils-error-to-json].
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: `@stdlib/utils/error-to-json`

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+var err1 = new SyntaxError( 'bad syntax' );
+// returns <SyntaxError>
+
+var json = err2json( err1 );
+/* returns
+    {
+        "type": "SyntaxError",
+        "name": "SyntaxError",
+        "message": "bad syntax",
+        "stack": "<stack>"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: s/\<stack\>/.../

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    }
+*/
+
+var str = JSON.stringify( json );
+// returns '{"type":"SyntaxError","name":"SyntaxError","message":"bad syntax","stack":"<stack>"}'
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: s/\<stack\>/.../

File: lib/node_modules/@stdlib/utils/error-reviver/README.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+</section>
+
+<!-- /.references -->
+
+<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="links">
+
+
+[utils-error-to-json]: https://github.com/kgryte/utils-error-to-json
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/README.md]: `@stdlib/utils/error-to-json`

File: lib/node_modules/@stdlib/utils/error-reviver/examples/index.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+var err1 = new SyntaxError( 'bad syntax' );
+// returns <SyntaxError>
+
+var json = err2json( err1 );
+/* returns
+    {
+        "type": "SyntaxError",
+        "name": "SyntaxError",
+        "message": "bad syntax",
+        "stack": "<stack>"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/examples/index.js]: s/\<stack\>/.../

File: lib/node_modules/@stdlib/utils/error-reviver/examples/index.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    }
+*/
+
+var str = JSON.stringify( json );
+// returns '{"type":"SyntaxError","name":"SyntaxError","message":"bad syntax","stack":"<stack>"}'
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/examples/index.js]: s/\<stack\>/.../

File: lib/node_modules/@stdlib/utils/error-reviver/test/test.ctors.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+var tape = require( 'tape' );
+var getKeys = require( 'object-keys' ).shim();
+var isPlainObject = require( '@stdlib/utils/is-plain-object' );
+var isFunction = require( '@stdlib/utils/is-function' );
+var ctors = require( './../lib/ctors.js' );
+
+
+// TESTS //
+
+tape( 'export is a function hash', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/test/test.ctors.js]: > main export is a...

File: lib/node_modules/@stdlib/utils/error-reviver/test/test.ctors.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+	var keys;
+	var i;
+
+	t.ok( true, __filename );
+	t.equal( isPlainObject( ctors ), true, 'export is an object' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/test/test.ctors.js]: > main export is an object

File: lib/node_modules/@stdlib/utils/error-reviver/test/test.js [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+		json = setup( types[ i ], types[ i ], msgs[i], 'boop'+i );
+
+		expected = new ctors[ i ]( msgs[i] );
+		expected.stack = 'boop' + i;
+
+		actual = JSON.parse( JSON.stringify( json ), reviveError );
+
+		t.ok( actual instanceof ctors[ i ], 'instance of type ' + types[i] );
+		t.equal( actual.message, expected.message, 'equal messages' );
+		t.equal( actual.stack, expected.stack );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/test/test.js]: Missing test message.

File: lib/node_modules/@stdlib/utils/error-reviver/test/test.js [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+	expected = {
+		'beep': 'boop'
+	};
+	actual = JSON.parse( '{"beep":"boop"}', reviveError );
+
+	t.deepEqual( actual, expected, 'deep equal' );
+
+	// Null edge case:
+	actual = JSON.parse( 'null', reviveError );
+	t.equal( actual, null );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/test/test.js]: Missing test message.

File: lib/node_modules/@stdlib/utils/error-reviver/test/test.js [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+	json.arr = [1,2,3,[4,5]];
+
+	err = JSON.parse( JSON.stringify( json ), reviveError );
+
+	t.equal( err.beep, json.beep, 'shallow properties' );
+
+	t.notEqual( err.arr, json.arr, 'separate instances' );
+	t.deepEqual( err.arr, json.arr, 'deep equal' );
+
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/utils/error-reviver/test/test.js]: Remove extra newline.

**Review by kgryte — CHANGES_REQUESTED** [review]

**Comment by Planeshifter** [comment]
Thanks for the comments! Fixed the issues.

**Review by kgryte — APPROVED** [review]

**Pull request merged**